ir/src/ir/operation.rs

```diff
@@ -1,22 +1,73 @@
+use value::SpannedMirValue;
+
 use crate::graph::NodeIndex;
 
 use super::*;
 
 /// [Operation] defines the various node types represented
-/// in the [AlgebraicGraph].
-#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+/// in the [MIR].
+#[derive(Debug, PartialEq, Eq, Clone)]
 pub enum Operation {
-    /// Evaluates to a [Value]
-    ///
-    /// This is always a leaf node in the graph.
-    Value(Value),
+    /// Begin primitive operations
+
+    /// Evaluates to a [TypedValue]
+    Value(SpannedMirValue),
     /// Evaluates by addition over two operands (given as nodes in the graph)
     Add(NodeIndex, NodeIndex),
     /// Evaluates by subtraction over two operands (given as nodes in the graph)
     Sub(NodeIndex, NodeIndex),
     /// Evaluates by multiplication over two operands (given as nodes in the graph)
     Mul(NodeIndex, NodeIndex),
+    /// Enforces a constraint (a given value equals Zero)
+    Enf(NodeIndex),
+
+    /// Begin structured operations
+    /// Call (func, arguments)
+    Call(NodeIndex, Vec<NodeIndex>),
+    /// Fold an Iterator according to a given FoldOperator and a given initial value
+    Fold(NodeIndex, FoldOperator, NodeIndex),
+    /// For (iterators, body, selector)
+    For(Vec<NodeIndex>, NodeIndex, Option<NodeIndex>),
+    /// If (condition, then, else)
+    If(NodeIndex, NodeIndex, NodeIndex),
+
+    /// A reference to a specific variable in a function
+    /// Variable(MirType, argument position)
+    Variable(SpannedVariable),
```

Comment:
Variable is improperly named here, it is meant to represent function parameters and not variables, will rename to Parameter to avoid further confusion

Reply:
So I think my actual question is: does `Variable` (or `Parameter` if you rename it) represent a named binding, or an SSA value?

My concern is that named bindings remain in the IR in any form. In other words, we shouldn't need to concern ourselves with names and their associated lexical scoping rules - particulary in the presence of uniquing. I would expect the IR to be composed of, essentially, blocks (with optional parameter lists) and operations (with operands and results), where block parameters, operands, and results are all SSA values. The dataflow of the program is then implicit in the edges of the value dependency graph.

As an aside, it might be a bit of a misnomer to call them SSA values, as AirScript doesn't _really_ need to enforce the dominance property of SSA-form programs (i.e. all uses of a value are dominated by the definition of that value). The only property we would need to enforce is that the value dependency graph doesn't have cycles. In effect, by construction, the program will be in SSA form, even though we don't really have a control flow graph per se, more of a loose scheduling order.

In any case, renaming from `Variable` to something else is probably still a good idea, since it implies the notion of a mutable binding, but I wanted to try and understand the distinction between these and `Value`. If I understand correctly, you are using `Variable` and `Value` to distinguish between what I call `BlockArgument` and `OpResult` in my other comment, is that correct?